api/turing/api/deployment_controller.go

```diff
@@ -231,6 +235,65 @@ func (c RouterDeploymentController) deployRouterVersion(
 		}
 	}
 
+	// Retrieve pyfunc ensembler if pyfunc ensembler is specified
+	if routerVersion.Ensembler != nil && routerVersion.Ensembler.Type == models.EnsemblerPyFuncType {
+		ensembler, err := c.EnsemblersService.FindByID(
+			*routerVersion.Ensembler.PyFuncRefConfig.EnsemblerID,
+			service.EnsemblersFindByIDOptions{
+				ProjectID: routerVersion.Ensembler.PyFuncRefConfig.ProjectID,
+			})
+		if err != nil {
+			return "", fmt.Errorf("failed to find specified ensembler: %w", err)
+		}
+
+		pyFuncEnsembler := ensembler.(*models.PyFuncEnsembler)
+
+		// Build image corresponding to the retrieved ensembler
+		request := imagebuilder.BuildImageRequest{
+			ProjectName: project.Name,
+			// Create a unique resource name that contains partial versioning (part of the run_id hash)
+			ResourceName: pyFuncEnsembler.Name + "-" + pyFuncEnsembler.RunID[:5],
```

Comment:
This has the side effect of ensuring that whenever a pyfunc ensembler gets updated separately by a user, redeploying a router would ensure that the new pyfunc ensembler gets rebuilt as part of the new web ensembler service image, instead of relying on a cached version. The arbitrary slice of the original hash is done due to constraints in name length.

Reply:
It's an arbitrary number to pick a subset of the entire RunID because the the RunID is pretty long. 

At 32 characters, we'll end up using more than half of the max character length for pod names (63) if we were to specify the entire RunID in the pod name of the image builder, leaving out less than half of the characters available for other information that can go into the name, i.e. job type, project name, ensembler name, ensembler id, the random 5-char alphanumeric string that k8s adds to pods, and the dashes that go in between these stuff.

Right now the naming convention for image building jobs go like `"service-builder-[projectName]-[ensemblerName]-[5 chars of the RunID]-[ensemblerID]-[k8s string]"`, but I think to make things simpler and shorter, maybe I'll throw out elements such as ensembler name and shorten the `service-builder` part. If we want to cut things further I'll have to insert these data as annotations into the pod. Not sure if we're sacrificing readability if we go that direction.

Reply:
> Not sure if we're sacrificing readability if we go that direction.

Yes, that would be sacrificing the readability which I agree is not such a good idea. Let's stick to the current implementation. Could you add a comment that the `[:5]` is arbitrary?

Also:
* Should Run ID go to the end of the name, to also be consistent with the batch builder naming?
* Can we drop the `builder-` part of the name, which doesn't seem to add too much value and can help shorten the name a bit?

Reply:
> Could you add a comment that the [:5] is arbitrary?

Yep sure.

> Can we drop the builder- part of the name, which doesn't seem to add too much value and can help shorten the name a bit?

That's true, besides, the namespace where the builder jobs get launched in is `image-builder` anyway so another `-builder-` is probably redundant.

> Should Run ID go to the end of the name, to also be consistent with the batch builder naming?

Hmm okay. I think to bypass the entire pod naming length issue altogether, I'll leave the pod naming conventions as they are without the Run ID, i.e. keeping `"service-builder-[projectName]-[ensemblerName]-[ensemblerID]-[k8s string]"` and `"job-builder-[projectName]-[ensemblerName]-[ensemblerID]-[k8s string]"`. On the other hand, I'll only inject the _entire_ Run ID into the docker image name (the one generated by the builder) instead, which will end up looking something like `"[registryName]/[projectName]-[ensemblerName]-[RunID]-service"` and `"[registryName]/[projectName]-[ensemblerName]-job"` respectively instead. Since docker image names have a way longer char limit (128), the Run ID's length isn't going to pose any problem. 

To do this I'll probably refactor the `BuildImageRequest` struct a little to carry an additional field `RunID`, which will be the `RunID` associated with the particular pyfunc ensembler.

Reply:
While in this case slicing a string may seem ok since we know that RunID falls within the ASCII range, if not it's better to use a Rune slice instead (https://go.dev/blog/strings).

Reply:
Hmm okay I think I might've created this tiny confusion over pod/image names in this thread. For pod names, there should no longer be a concern over name lengths because I'm planning to remove RunID from them altogether. 

Furthermore, the unique names concern only exists for image names because that's how Turing API checks for images that have already been created to avoid recreating an image with the same ensembler. To address this, I'll be adding the entire RunID to the image name so that serves to distinguish between images built with ensemblers with the same ensembler name, ensemblerID, projectID but with different versions (because a user can update an ensembler with a different ensembling logic). 

The confusion/problem arose when I was initially trying to splice the RunID with the ensembler name as a *new* ensembler name used by the existing name generators, hence confounding both naming concerns together. This was initially done to avoid having to refactor the existing `imageBuilder` struct since the original batch ensembling images don't have RunIDs in their image names to distinguish themselves (i.e. if I understood the existing implementation correctly, a user rebuilding a batch ensembling job with an updated ensembler would probably not have a new image rebuilt because an image containing the very same project and ensembler name already exist in the registry, leading to the cached image being used for the ensembling job, which would be incorrect). 

Okay apologies for the really long comment 😅  but I hope this clears things up!